mongo/writeconcern/writeconcern.go

```diff
@@ -17,34 +21,126 @@ import (
+	// mongod instances or to mongod instances with specified tags. It sets the the "w" option
+	// in a MongoDB write concern.
+	//
+	// W must be a type string or int value.
+	//
+	// For more information about the "w" option, see
+	// https://www.mongodb.com/docs/manual/reference/write-concern/#w-option
+	W interface{}
+
+	// Journal requests acknowledgment from MongoDB that the write operation has been written to the
+	// on-disk journal. It sets the "j" option in a MongoDB write concern.
+	//
+	// For more information about the "j" option, see
+	// https://www.mongodb.com/docs/manual/reference/write-concern/#j-option
+	Journal *bool
+
+	// WTimeout specifies a time limit for the write concern. It sets the "wtimeout" option in a
+	// MongoDB write concern.
+	//
+	// It is only applicable for "w" values greater than 1. Using a WTimeout and setting Timeout on
+	// the Client at the same time will result in undefined behavior.
+	//
+	// For more information about the "wtimeout" option, see
+	// https://www.mongodb.com/docs/manual/reference/write-concern/#wtimeout
+	WTimeout time.Duration
+}
 
-	// NOTE(benjirewis): wTimeout will be deprecated in a future release. The more general Timeout
-	// option may be used in its place to control the amount of time that a single operation can run
-	// before returning an error. Using wTimeout and setting Timeout on the client will result in
-	// undefined behavior.
-	wTimeout time.Duration
+// Majority returns a WriteConcern that requests acknowledgment that write operations have been
+// durably committed to the calculated majority of the data-bearing voting members. If journaling is
+// enabled on mongod, nodes only acknowledge when the write operation has been written to the
+// on-disk journal.
+//
+// For more information about write concern "w: majority", see
+// https://www.mongodb.com/docs/manual/reference/write-concern/#mongodb-writeconcern-writeconcern.-majority-
+func Majority() *WriteConcern {
+	return &WriteConcern{W: "majority"}
+}
+
+// Custom returns a WriteConcern that requests acknowledgment that the write operation has
+// propagated to tagged members that satisfy the custom write concern defined in
+// "settings.getLastErrorModes". If journaling is enabled on mongod, nodes only acknowledge when the
+// write operation has been written to the on-disk journal.
+//
+// For more information about custom write concern names, see
+// https://www.mongodb.com/docs/manual/reference/write-concern/#mongodb-writeconcern-writeconcern.-custom-write-concern-name-
+func Custom(tag string) *WriteConcern {
+	return &WriteConcern{W: tag}
+}
+
+// W0 returns a WriteConcern that requests no acknowledgment of the write operation.
+//
+// For more information about write concern "w: 0", see
+// https://www.mongodb.com/docs/manual/reference/write-concern/#mongodb-writeconcern-writeconcern.-number-
+func W0() *WriteConcern {
```

Comment:
The main benefit is that `{"w": "majority"}`, `{"w": 0}`, and `{"w": 1}` are the most commonly used write concerns, so we want to provide users a quick way to define those.

Defining a convenience function like
```go
func W(int) *WriteConcern
```
that lets you specify any numeric `w` like
```go
writeconcern.W(1)
```
could also make sense, but there is already a package-level `W` function, so that's not possible currently.

Another alternative is to provide convenience functions that describe what behavior the write concern causes, like
```go
func Unacknowledged() *WriteConcern // Returns {"w": 0}
func Acknowledged() *WriteConcern   // Returns {"w": 1}
```
However, we can't extend that pattern to `Majority` because that name is also already taken.

TLDR: I picked `W0` and `W1` partly because they're common write concerns and partly to avoid naming collisions.